sf-stream/src/runtime.rs

```diff
@@ -109,125 +92,162 @@ impl SfStreamer {
+                    }
+                    if curr_version != txn_version {
+                        error!(
+                            "Block {} failed validation: missing version {}",
+                            self.current_block_height, curr_version,
+                        );
+                        sleep(Duration::from_millis(500)).await;
+                        return vec![];
+                    }
                     let txn_proto = self.convert_transaction(txn);
-                    self.print_transaction(&txn_proto).await;
+                    self.print_transaction(&txn_proto);
                     result.push(txn_proto);
-                    self.current_version = txn_version;
+                    curr_version += 1;
+                }
+                if curr_version - 1 != block_last_version {
+                    error!(
+                        "Block {} failed validation: last version supposed to be {} but getting {}",
+                        self.current_block_height,
+                        block_last_version,
+                        curr_version - 1,
+                    );
+                    sleep(Duration::from_millis(500)).await;
+                    return vec![];
                 }
-                self.current_version += 1;
             }
             Err(err) => {
                 error!("[sf-stream] failed to get transactions: {}", err);
                 sleep(Duration::from_millis(100)).await;
                 return vec![];
             }
         }
+        println!("FIRE BLOCK_END {}", self.current_block_height);
+        metrics::BLOCKS_SENT.inc();
+        self.current_block_height += 1;
         result
     }
 
-    pub fn maybe_update_from_block_metadata(&mut self, transaction: &Transaction) {
-        if let Transaction::BlockMetadataTransaction(bmt) = transaction {
-            // TODO: ADD BLOCK HEIGHT UPDATES ONCE ITS IN BMT
-            self.block_height += 1;
-            self.current_epoch = bmt.epoch.0
-        }
+    pub fn convert_transaction(&self, transaction: Transaction) -> TransactionPB {
+        convert_transaction(&transaction, self.current_block_height, self.current_epoch)
     }
 
-    pub fn convert_transaction(&mut self, transaction: Transaction) -> extractor::Transaction {
-        self.maybe_update_from_block_metadata(&transaction);
-        convert_transaction(&transaction, self.block_height, self.current_epoch)
+    /// First, and only first, transaction in a block has to be bmt or genesis
+    fn validate_transaction_type(&self, is_first_txn: bool, transaction: &Transaction) -> bool {
+        let is_bm_or_genesis = match transaction {
+            Transaction::BlockMetadataTransaction(_) => true,
+            Transaction::GenesisTransaction(_) => true,
+            Transaction::UserTransaction(_) => false,
+            Transaction::StateCheckpointTransaction(_) => false,
+            Transaction::PendingTransaction(_) => false,
+        };
```

Comment:
```
        let is_bm_or_genesis = match transaction {
            Transaction::BlockMetadataTransaction(_) => true,
            Transaction::GenesisTransaction(_) => true,
            _ => false,
        };
```

Reply:
ah love this.

Reply:
xclippy gave me something even better lol 
```
matches!(
            transaction,
            Transaction::BlockMetadataTransaction(_) | Transaction::GenesisTransaction(_)
        );
```

Reply:
nice!